br/pkg/restore/data.go

```diff
@@ -0,0 +1,503 @@
+					return errors.Trace(err)
+				}
+			}
+
+			metaChan <- tikvMeta
+			return nil
+		})
+	}
+
+	for i := 0; i < totalTiKVs; i++ {
+		tikvMeta := <-metaChan
+		recovery.regionMetas[i] = tikvMeta
+		log.Debug("TiKV", zap.Int("recived from tikv", int(tikvMeta.storeId)))
+	}
+
+	return eg.Wait()
+}
+
+func (recovery *Recovery) getTotalRegions() int {
+	// Group region peer info by region id.
+	var regions = make(map[uint64][]Regions, 0)
+	for _, v := range recovery.regionMetas {
+		storeId := v.storeId
+		for _, m := range v.recoveryMeta {
+			if regions[m.RegionId] == nil {
+				regions[m.RegionId] = make([]Regions, 0, recovery.totalStores)
+			}
+			regions[m.RegionId] = append(regions[m.RegionId], Regions{m, storeId})
+		}
+	}
+	return len(regions)
+}
+
+// TODO: function provide for dump server received peer info into a file
+// function shall be enabled by debug enabled
+func (recovery *Recovery) dumpRegionInfo() {
+	log.Debug("dump region info")
+	// Group region peer info by region id.
+	var regions = make(map[uint64][]Regions, 0)
+	for _, v := range recovery.regionMetas {
+		storeId := v.storeId
+		for _, m := range v.recoveryMeta {
+			if regions[m.RegionId] == nil {
+				regions[m.RegionId] = make([]Regions, 0, recovery.totalStores)
+			}
+			regions[m.RegionId] = append(regions[m.RegionId], Regions{m, storeId})
+		}
+	}
+
+	for regionID, peers := range regions {
+		log.Debug("Region", zap.Uint64("RegionID", regionID))
+		for _, m := range peers {
+			log.Debug("tikv", zap.Int("storeId", int(m.storeId)))
+			log.Debug("meta:", zap.Uint64("last_log_term", m.GetLastLogTerm()))
+			log.Debug("meta:", zap.Uint64("last_index", m.GetLastIndex()))
+			log.Debug("meta:", zap.Uint64("version", m.GetVersion()))
+			log.Debug("meta:", zap.Bool("tombstone", m.GetTombstone()))
+			log.Debug("meta:", zap.ByteString("start_key", m.GetStartKey()))
+			log.Debug("meta:", zap.ByteString("end_key", m.GetEndKey()))
+		}
+	}
+
+}
+
+// TODO : function provide for dump RecoveryPlan
+func (recovery *Recovery) dumpRecoverPlan() {
+	log.Debug("dump recovery plan")
+	// Group region peer info by region id.
+	for storeID, _ := range recovery.recoveryPlan {
+		log.Debug("TiKV", zap.Uint64("store id", storeID))
+	}
+}
+
+// send the recovery plan to recovery region (force leader etc)
+func (recovery *Recovery) RecoverRegions(ctx context.Context, allStores []*metapb.Store) (err error) {
+	eg, ectx := errgroup.WithContext(ctx)
+	totalTiKVs := len(recovery.recoveryPlan)
+	workers := utils.NewWorkerPool(uint(totalTiKVs), "Recover Regions")
+
+	for storeId, plan := range recovery.recoveryPlan {
+		storeAddr := getStoreAddress(allStores, storeId)
+		tikvClient, err := recovery.newTiKVRecoveryClient(ectx, storeAddr)
+		if err != nil {
+			log.Error("create tikv client failed", zap.Uint64("storeID", storeId))
+			return errors.Trace(err)
+		}
+		cmd := plan
+		workers.ApplyWithIDInErrorGroup(eg, func(id uint64) error {
+			stream, err := tikvClient.RecoverCmd(ectx)
```

Comment:
RecoverCmd is not a clear name, do we have chance to rename it to `RecoverRegions` of `RecoverRegionMetas`?